src/Bridge/Doctrine/Orm/SubcollectionDataProvider.php

```diff
@@ -0,0 +1,140 @@
+
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Extension\QueryResultCollectionExtensionInterface;
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Util\IdentifierManagerTrait;
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Util\QueryNameGenerator;
+use ApiPlatform\Core\DataProvider\SubcollectionDataProviderInterface;
+use ApiPlatform\Core\Exception\ResourceClassNotSupportedException;
+use ApiPlatform\Core\Exception\RuntimeException;
+use ApiPlatform\Core\Metadata\Property\Factory\PropertyMetadataFactoryInterface;
+use ApiPlatform\Core\Metadata\Property\Factory\PropertyNameCollectionFactoryInterface;
+use Doctrine\Common\Persistence\ManagerRegistry;
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\Mapping\ClassMetadataInfo;
+
+/**
+ * Subcollection data provider for the Doctrine ORM.
+ *
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+class SubcollectionDataProvider implements SubcollectionDataProviderInterface
+{
+    use IdentifierManagerTrait;
+
+    private $managerRegistry;
+    private $collectionExtensions;
+
+    /**
+     * @param ManagerRegistry                        $managerRegistry
+     * @param PropertyNameCollectionFactoryInterface $propertyNameCollectionFactory
+     * @param PropertyMetadataFactoryInterface       $propertyMetadataFactory
+     * @param QueryItemExtensionInterface[]          $itemExtensions
+     */
+    public function __construct(ManagerRegistry $managerRegistry, PropertyNameCollectionFactoryInterface $propertyNameCollectionFactory, PropertyMetadataFactoryInterface $propertyMetadataFactory, array $collectionExtensions = [])
+    {
+        $this->managerRegistry = $managerRegistry;
+        $this->propertyNameCollectionFactory = $propertyNameCollectionFactory;
+        $this->propertyMetadataFactory = $propertyMetadataFactory;
+        $this->collectionExtensions = $collectionExtensions;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws RuntimeException
+     */
+    public function getSubcollection(string $resourceClass, array $identifiers, array $context, string $operationName = null)
+    {
+        $manager = $this->managerRegistry->getManagerForClass($resourceClass);
+        if (null === $manager) {
+            throw new ResourceClassNotSupportedException();
+        }
+
+        $repository = $manager->getRepository($resourceClass);
+        if (!method_exists($repository, 'createQueryBuilder')) {
+            throw new RuntimeException('The repository class must have a "createQueryBuilder" method.');
+        }
+
+        if (!isset($context['identifiers']) || !isset($context['property'])) {
+            throw new ResourceClassNotSupportedException('The given resource class is not a subcollection.');
+        }
+
+        $queryBuilder = $repository->createQueryBuilder('o');
+        $queryNameGenerator = new QueryNameGenerator();
+        $previousQueryBuilder = null;
+        $previousAlias = null;
+
+        $num = count($context['identifiers']);
+
+        while ($num--) {
```

Comment:
Because of the while ? :smile: 

It is complicated because we're doing a recursive query where each previous part is included in the next one:

```
$previousQueryBuilder->andWhere($qb->expr()->in($previousAlias, $qb->getDQL()));
```

and 

```
$queryBuilder->where(
    $queryBuilder->expr()->in('o', $previousQueryBuilder->getDQL())
);
```

The rest is pretty straightforward, we first select the data we need, than we're adding identifiers to the where clause.